### ISSUE TITLE
Send message on event

### ISSUE THREAD

**Issue by ricksancho** [go-chat-bot/bot#37]
I currently developing a bot that monitor a websocket and forward messages received to a channel. Your library looks great to does this task and will allow to add more feature but the issue I get with is I don't find how to send a message without being into the context of a command most of the method to do that are a not exported. I would like to know if you have any idea how I can achieve this into a plugin?

**Comment by fabioxgn** [comment]
Hello @ricksancho, thanks for the interest in this project.

Right now this is not supported, I'll think about the best way to implement that in the near future.

I thought about tricking the bot to execute your plugin using the `MessageReceived` method, something like:

- Write a plugin that responds to a certain command, ex: `!forward_msg`
- Write your websocket handler
- Call `botInstance.MessageReceived("#channel", "!forward_msg your message", &bot.User{Nick: "Sender Nickname"})`

But that would not work, as the bot instance is private in the protocol implementation.

**Comment by ricksancho** [comment]
Thanks @fabioxgn, call on MessageReceived with some kind of echo plugin was my first idea, but I come to same conclusion, that would not work due to private instance. I was lacking of idea too on how to patch this nicely and just finish by creating a new bot for this task only (just fwd the websocket message to channel with an irc lib).
If I have spare time I will try to come up with a solution.

**Comment by doenietzomoeilijk** [comment]
I'm actually interested in this as well - if I understand things correctly this would also enable cross-network chatter, ie, have a Telegram bot notify me on certain events / commands from IRC. I suppose you could run several networks in one bot using goroutines and have a channel or mutex for cross-talk, but I'm not exactly sure. I'm pretty new to Go, so there could be a better / more elegant way to handle it.

Either way, in both cases it's a matter of "injecting events in a running stream", right? If not, I'll open a separate issue.

**Comment by fabioxgn** [comment]
@doenietzomoeilijk yes, somehow it would, but right now each protocol runs on a different bot instance, to to cross chat I think that the easiest way is to make some events public on the bot and then write some kind of router which will receive all the messages from all instances and forward them to the other instances.

It is not impossible to do but will require some work.

**Comment by doenietzomoeilijk** [comment]
@fabioxgn The way I was thinking of - and I'm in no way sure that it's possible, or the best way - was using several goroutines in one bot instance, like so:

```go
func main() {
	go slack.Run("my-slack-token")
	go telegram.Run("my-telegram-token")
	go irc.Run(myIrcConfig)
}
```

Would that work, or does that create more problems than it solves? Adding a channel argument to the respective Run methods would allow you to use one channel across all networks, I think, similar to [this example](https://gobyexample.com/worker-pools).

Just thinking out loud, and as I said, I'm new to Go so I could very well be missing some obvious things. :)

**Comment by fabioxgn** [comment]
@doenietzomoeilijk that will work, it is exactly like that that I run my own bot in production, as you can see here:

https://github.com/fabioxgn/go-bot-heroku/blob/master/main.go#L35

You just can't run the last one in a goroutine or the program will exist, either you wait something or run one of them in the main loop as I do.

**Comment by doenietzomoeilijk** [comment]
@fabioxgn Ah, good to know I wasn't completely off, then. So would creating a channel before starting all the bots, and then providing that as an extra argument to all of them work? Can you "use" a channel from a "regular" routine, in your case the Slack one?

**Comment by fabioxgn** [comment]
@doenietzomoeilijk yes, it will work, btw that's the way channels are ment to be used, see: https://gobyexample.com/channels

**Comment by doenietzomoeilijk** [comment]
@fabioxgn Excellent, so then all we need is a mechanism for one bot runner to send data to another bot runner, either directed (an IRC bot that explicitely targets the Slack bot running in the same instance) or as a general queue (an IRC bot that stuffs something into a "cross-instance" channel, for the other channels to maybe act upon). However, that requires each bot to know what type it is, and to make that known to plugins so they can act on certain actions on certain network types, but not on others. Alternatively, you'd have to be able to hook up certain plugins to certain bot only, so for example my IRC bot gets catgifs and chucknorris, and my Slack bot gets chucknorris and godoc. Either option has pros and cons, what's your thoughts?

**Comment by fabioxgn** [comment]
@doenietzomoeilijk The most complicated part is which channel to send the messages to, I think that the implementation will have to solve this either fixing a channel on startup on dispatching a message to the other protocol and the receiver will have to know in which channel to post to.

**Comment by doenietzomoeilijk** [comment]
@fabioxgn Channel as in chat channel or as in go communication channel?

**Comment by fabioxgn** [comment]
@doenietzomoeilijk chat channel.

**Comment by doenietzomoeilijk** [comment]
OK, gotcha. And yes, you need to know which bot+channel you need to send to. I'm currently somewhat running into this with the quotes-script - I want to keep all the quotes in one database, but keep all the channels apart. That'd be a tad easier if, at command time, I had at least the network type (irc|slack|telegram), and possibly even more info (if I have two IRC bots running I still want to be able to tell apart the two connections).

How open are you to adding more info to the `Cmd` struct? (And on that note, what's `struct incomingMessage` for? I don't see it used anywhere?)

**Comment by fabioxgn** [comment]
@doenietzomoeilijk maybe the correct way to handle this is instead of running multiple bot instances, allow one instance to run several protocols, this will require some major refactorings in the end I think it's the correct way to do.

I run mutiple instances of the bot as it was the easiest thing to do, but when you start adding complexity I think that it's better to do it correctly now than regret it later.

About the `incomingMessage` I think it is some leftover from a previous version, it is not used.

**Comment by doenietzomoeilijk** [comment]
I'll see what I can come up with in terms of ideas. For now my bot theoretically is limited to IRC anyway, so I want to focus on getting that done; refactorings and multi-network can come after that.

**Comment by fabioxgn** [comment]
I'm closing this issue as I think there's nothing to be done for now.
